fdbserver/workloads/TenantManagementWorkload.actor.cpp

```diff
@@ -626,6 +667,45 @@ struct TenantManagementWorkload : TestWorkload {
 					ASSERT(operationType == OperationType::METACLUSTER != self->useMetacluster);
 					return Void();
 				} else if (e.code() == error_code_cluster_no_capacity) {
+					int64_t lastTenantId = wait(getLastTenantId(self->dataDb));
+					int tenantsCreating = newTenants;
+					// For special keys and management transactions since we create multiple tenants it's possible some
+					// of them are duplicates in which case the whole transaction will fail if we overshoot capacity but
+					// newTenants will not accurately reflect the total ids we attempt to assign
+					if (operationType == OperationType::SPECIAL_KEYS ||
+					    operationType == OperationType::MANAGEMENT_TRANSACTION) {
+						tenantsCreating = tenantsToCreate.size();
+					}
+					TraceEvent(SevWarn, "TenantCapcityLimitsExceeded")
+					    .detail("LastTenantId", lastTenantId)
+					    .detail("NewTenants", tenantsCreating)
+					    .detail("ClientID", self->clientId)
+					    .detail("Opetation", operationType);
+					// Overshot our capacity because we cannot assign anymore tenant ids with the same prefix
+					if (TenantAPI::getTenantIdPrefix(lastTenantId) !=
+					    TenantAPI::getTenantIdPrefix(lastTenantId + tenantsCreating)) {
+						CODE_PROBE(true, "prefix change from reaching tenant capacity limits");
+						ASSERT(TenantAPI::getTenantIdPrefix(lastTenantId) == self->tenantIdPrefix);
+
+						// It's possible that we created tenants but the createTenantImpl actor timed out so we had to
+						// retry. Between when this retry is invoked to when the attempt is made to insert tenants the
+						// tenant id could have increased greatly throwing capacity errors even though we're not
+						// inserting new tenants. Thus when we get to this codepath we need to ensure we update our
+						// local state and add any missing tenants that were created.
+						state std::map<TenantName, TenantMapEntry>::iterator curTenantItr;
+						for (curTenantItr = tenantsToCreate.begin(); curTenantItr != tenantsToCreate.end();
+						     ++curTenantItr) {
+							state Optional<TenantMapEntry> curEntry =
+							    wait(self->tryGetTenant(curTenantItr->first, operationType));
+							if (curEntry.present()) {
+								wait(updateLocalState(self, curTenantItr, operationType, originalReadVersion));
+							} else {
+								break;
+							}
+						}
+
+						return Void();
```

Comment:
I think this test works, though it's also a lot of code to test what is essentially meant to be thought of as an ASSERT that we shouldn't be hitting. I wonder if there's a way we could test this with a more targeted test, such as some sort of unit test or a very simple standalone test. For example, we could create a metacluster (or standalone cluster), explicitly set the last tenant ID to something at or near maximum, and then verify that we cannot create tenants.

Reply:
I would argue that doing it this way allows us to be more confident in our change and technically if we do ever hit the limit in our test (which is theoretically possible although highly unlikely) it would be nicer to handle it

Reply:
My main concern is that this workload is already quite complex, and I don't want to add too much more complexity to test an assertion that should never be hit. I think writing a simple test that just checks that the limit works helps to keep this particular test from doing too much.

If this test is ever hitting the limit without the code that forces tenant IDs to advance by your new huge number, then we should adjust the test to avoid advancing the IDs so fast.

Reply:
Sure i guess i can avoid testing it here then